base/usr/share/man/man1/grep.1
.TH GREP 1 "March, 2026"

.SH NAME
grep \- Search for patterns in files.

.SH SYNOPSIS
\fBgrep\fP  [\fIOPTION\fP] \fIPATTERN\fP [\fIFILE\fP...]
.PP
\fBfgrep\fP [\fIOPTION\fP] \fISTRING\fP [\fIFILE\fP...]

.SH DESCRIPTION
Search for \fIPATTERN\fP in each file, or the standard input.
.PP
This \fIgrep\fP has a bizarre, non-standard, and rather broken
regex engine and probably doesn't do what you want for anything
more complex than a simple matching pattern. It is not at all
POSIX\-compliant.
.TP
.B \-c
Instead of printing matches, print counts of matched lines.
.TP
.B \-i
Ignore case in input and pattern.
.TP
.B \-l
Instead of printing matches, print the names of files that match.
This takes precedence over most other options.
.TP
.B \-n
Prefix each printed match with the line number it appeared on.
.TP
.B \-o
Print only the matching parts of a line, separating each match
with a line feed.
.TP
.B \-q
Exit immediately with a \fB0\fP exit code when a match (or, with \fB\-v\fP,
a non-match) is found, without printing any matches.
.TP
.B \-v
Invert matching: instead of printing lines that match the pattern,
prints lines that do not match the pattern.
.TP
.B \-x
Require that \fIPATTERN\fP match the whole line to count as a match.
.TP
.B \-F
Treat \fIPATTERN\fP as a fixed string. This is the default mode when
invoked as \fIfgrep\fP.
.TP
.B \-A \fIn\fP
Print up to \fIn\fP lines of context after a match.
Groups of matches will be separated by "\-\-", unless they merge.
.TP
.B \-B \fIn\fP
Print up to \fIn\fP lines of context before a match.
Groups of matches will be separated by "\-\-", unless they merge.
.TP
.B \-C \fIn\fP
Combined the \fB\-A\fP and \fB\-B\fP options to print up to \fIn\fP
lines of context both before and after a match. You may also specify
\fB\-A\fP and \fB\-B\fP separately to get different context on
either side of a match. As with the other options,
groups of matches will be separated by "\-\-", unless they merge.
.TP
.B \-\-color
Set the mode for displaying matches using coloring. With no additional
option, this enables color when the output is a tty and the matching part
of a line will be highlighted in red. This option can be configured as
follows:
.RS
.TP
.B \-\-color=auto
Enable color when the output is a TTY. This is the default.
.TP
.B \-\-color=never
Disable color even if the output is a TTY.
.TP
.B \-\-color=always
Always enable color, even if the output is not a TTY.
.RE

.SH PATTERNS
The pattern engine here was originally built for
.IR bim (1)
and was then adapted to be more like the one expected of a \fIgrep\fP.
.PP
The following subpatterns are supported to some extent:
.TP
.B [abc]
Matches one of a set of characters.
.TP
.B [^abc]
Matches one character not from the given set.
.TP
.B (abc)
Match a group. This does nothing useful, as groups can not be repeated in this
implementaiton.
.TP
.B .
Match any single character.
.TP
.B ^
Match the start of a line.
.TP
.B $
Match the end of a line.
.PP
Single characters and bracket expressions can be repeated as follows:
.TP
.B *
Match any number of occurances.
.TP
.B \e?
Match zero or one time.
.TP
.B \e+
Match at least one occurance.

.SH BUGS
 \- The regex engine is not Unicode-aware.
 \- Repetitions are not available on groups.

.SH SEE ALSO
.BR bim (1)

.SH AUTHOR
This \fIgrep\fP was written by K. Lange <klange@toaruos.org>



apps/grep.c
/**
 * @brief grep - almost acceptable grep
 *
 * Based on the regex search matcher in bim.
 *
 * @copyright
 * This file is part of ToaruOS and is released under the terms
 * of the NCSA / University of Illinois License - see LICENSE.md
 * Copyright (C) 2022-2026 K. Lange
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <ctype.h>
#include <errno.h>
#include <libgen.h>

static int invert = 0;
static int ignorecase = 0;
static int quiet = 0;
static int only_matching = 0;
static int counts = 0;
static int is_fgrep = 0;
static int whole_lines = 0;
static int list_files = 0;
static int line_numbers = 0;
static int use_color = 0;
static int suppress_errors = 0;
static size_t print_before = 0;
static size_t print_after = 0;

/**
 * This regex engine is adapted from bim.
 *
 * It has been modified to be more like POSIX BREs, but it's still pretty bad.
 *
 * While we can support * matches for square brackets, we don't support them
 * for groups. We don't support character classes (colons), repetition ranges,
 * and definitely not backrefs.
 */
struct MatchQualifier {
	int (*matchFunc)(struct MatchQualifier*,char,int);
	union {
		char matchChar;
		struct {
			char * start;
			char * end;
		} matchSquares;
	};
};

typedef enum {
	MATCH_OP_EQ,
	MATCH_OP_GE,
	MATCH_OP_LE,
} match_op;

/**
 * Helper for handling smart case sensitivity.
 */
static inline int match_char_internal(char a, char b, int ignorecase, match_op op) {
	if (ignorecase) {
		a = tolower(a);
		b = tolower(b);
	}
	switch (op) {
		/* Equality */
		case MATCH_OP_EQ: return a == b;
		case MATCH_OP_GE: return a >= b;
		case MATCH_OP_LE: return a <= b;
	}
	return 0;
}

/*
 * Basic single-character matcher.
 */
int match_char(struct MatchQualifier * self, char b, int mode) {
	return match_char_internal(self->matchChar, b, mode, MATCH_OP_EQ);
}

/*
 * Match collections of characters.
 */
int match_squares(struct MatchQualifier * self, char c, int mode) {
	char * start = self->matchSquares.start;
	char * end = self->matchSquares.end;
	char * t = start;
	int good = 1;
	if (*t == '^') { t++; good = 0; }
	while (t != end) {
		char test = *t++;
		if (test == '\\' && *t && strchr("\\]",*t)) {
			test = *t++;
		} else if (test == '\\' && *t == 't') {
			test = '\t'; t++;
		}

		if (*t == '-') {
			t++;
			if (t == end) return 0;
			char right = *t++;
			if (right == '\\' && *t && strchr("\\]",*t)) {
				right = *t++;
			} else if (right == '\\' && *t == 't') {
				right = '\t'; t++;
			}
			if (match_char_internal(c,test,mode,MATCH_OP_GE) && match_char_internal(c,right,mode,MATCH_OP_LE)) return good;
		} else {
			if (match_char_internal(c,test,mode,MATCH_OP_EQ)) return good;
		}
	}
	return !good;
}

/*
 * Match any single character.
 */
int match_dot(struct MatchQualifier * self, char c, int mode) {
	return 1;
}

struct Line {
	int actual;
	char * text;
};

int regex_matches(struct Line * line, int j, char * needle, int ignorecase, int *len, char **needleout) {
	int k = j;
	char * match = needle;
	if (*match == '^') {
		if (j != 0) return 0;
		match++;
	}
	while (k < line->actual + 1) {
		if (needleout && *match == ')') {
			*needleout = match + 1;
			if (len) *len = k - j;
			return 1;
		}
		if (*match == '\0') {
			if (needleout) return 0;
			if (len) *len = k - j;
			return 1;
		}
		if (*match == '$') {
			if (k != line->actual) return 0;
			match++;
			continue;
		}

		struct MatchQualifier matcher = {match_char, .matchChar=*match};
		if (*match == '.') {
			matcher.matchFunc = match_dot;
			match++;
		} else if (*match == '\\' && strchr("$^/\\.[]*()",match[1]) != NULL) {
			matcher.matchChar = match[1];
			match += 2;
		} else if (*match == '\\' && match[1] == 't') {
			matcher.matchChar = '\t';
			match += 2;
		} else if (*match == '[') {
			char * s = match+1;
			char * e = s;
			while (*e && *e != ']') {
				if (*e == '\\' && e[1] == ']') e++;
				e++;
			}
			if (!*e) break; /* fail match on unterminated [] sequence */
			match = e + 1;
			matcher.matchFunc = match_squares;
			matcher.matchSquares.start = s;
			matcher.matchSquares.end = e;
		} else if (*match == '(') {
			match++;
			int _len;
			char * newmatch;
			if (!regex_matches(line, k, match, ignorecase, &_len, &newmatch)) break;
			match = newmatch;
			k += _len;
			continue;
		} else {
			match++;
		}
		if (*match == '\\' && match[1] == '?') {
			/* Optional */
			match += 2;
			if (matcher.matchFunc(&matcher, line->text[k], ignorecase)) {
				int _len;
				if (regex_matches(line,k+1,match,ignorecase,&_len, needleout)) {
					if (len) *len = _len + k + 1 - j;
					return 1;
				}
			}
			continue;
		} else if ((*match == '\\' && match[1] == '+') || *match == '*') {
			if (*match == '\\' /* && match[1] == '+' */) {
				/* Must match at least one */
				match += 2;
				if (!matcher.matchFunc(&matcher, line->text[k], ignorecase)) break;
				k++;
			} else {
				match++;
			}

			int _j = k;
			while (_j < line->actual + 1) {
				if (_j < line->actual && !matcher.matchFunc(&matcher, line->text[_j], ignorecase)) break;
				_j++;
			}
			while (_j >= k) {
				int _len;
				if (regex_matches(line, _j, match, ignorecase, &_len, needleout)) {
					if (len) *len = _len + _j - j;
					return 1;
				}
				_j--;
			}
			return 0;
		} else {
			if (!matcher.matchFunc(&matcher, line->text[k], ignorecase)) break;
			k++;
		}
	}
	return 0;
}

/**
 * Determine if 'line' matches a particular pattern 'needle' using the current
 * global mode settings (is_fgrep, ignorecase), starting from a given line, and
 * store the resulting length of a match in 'len'.
 */
static int subsearch_matches(struct Line * line, int j, char * needle, int *len) {
	if (is_fgrep) {
		/* Does 'line' starting at 'j' match 'needle' */
		const char *n = needle;
		for (; *n; ++n, ++j) if (j >= line->actual || !match_char_internal(*n, line->text[j], ignorecase, MATCH_OP_EQ)) return 0;
		*len = n - needle;
		return 1;
	}
	return regex_matches(line, j, needle, ignorecase, len, NULL);
}

struct ContextLine {
	struct ContextLine * next;
	struct ContextLine * prev;
	char * buf;
	size_t len;
	const char * filename;
	int line;
};

static void print_dash(const char * dash, int nl) {
	printf("%s%s%s%s",
		use_color ? "\033[36m" : "",
		dash,
		use_color ? "\033[0m" : "",
		nl ? "\n" : "");
}

static void print_filename(const char * filename, char * sep) {
	printf("%s%s%s",
		use_color ? "\033[35m" : "",
		filename,
		use_color ? "\033[0m" : "");
	print_dash(sep, 0);
}

static void print_line_number(int line_number, char * sep) {
	printf("%s%d%s",
		use_color ? "\033[32m" : "",
		line_number,
		use_color ? "\033[0m" : "");
	print_dash(sep, 0);
}

static struct ContextLine * context_head = NULL;
static struct ContextLine * context_tail = NULL;
static size_t context_len = 0;
static int context_overflowed = 0;
static void dump_context(int printing) {
	struct ContextLine * ctx = context_tail;
	while (ctx) {
			if (printing && ctx != context_head) {
			if (ctx->filename) print_filename(ctx->filename, "-");
			if (ctx->line) print_line_number(ctx->line, "-");
			fprintf(stdout, "%.*s\n", (int)ctx->len, ctx->buf);
		}
		struct ContextLine * p = ctx->prev;
		free(ctx->buf);
		free(ctx);
		ctx = p;
	}
	context_head = NULL;
	context_tail = NULL;
	context_len = 0;
	context_overflowed = 0;
}

int usage(char ** argv) {
#define _I "\033[3m"
#define _E "\033[0m\n"
	fprintf(stderr, "usage: %s [-cilnoqsvxF] PATTERN [FILE...]\n"
		"\n"
		"Search for PATTERN in each file.\n"
		"Take care that this grep's pattern engine is limited and not POSIX-compliant.\n"
		"\n"
		" Supported options:\n"
		"  -c      " _I "Instead of printing matches, print counts of matched lines." _E
		"  -i      " _I "Ignore case in input and pattern." _E
		"  -l      " _I "Instead of printing matches, print the names of files that.\n"
		"          match. Takes precedence over most other options." _E
		"  -n      " _I "Prefix each printed match with the line number it appeared on." _E
		"  -o      " _I "Print only the matching parts of each line, separating\n"
		"          each match with a line feed." _E
		"  -q      " _I "Exit immediately with 0 when a match (or, with -v,\n"
		"          non-match) is found; do not print matches." _E
		"  -s      " _I "Suppress the output of errors that would normally go to stderr." _E
		"  -v      " _I "Invert match - print lines that do not match pattern." _E
		"  -x      " _I "PATTERN must match a whole line." _E
		"  -F      " _I "Treat PATTERN as a fixed string (acts as 'fgrep')." _E
		"  -A " _I " n   Print up to n non-matching lines after a match." _E
		"  -B " _I " n   Print up to n non-matching lines before a match." _E
		"  -C " _I " n   Print up to n non-matching lines each before and after a match." _E
		"  --help  " _I "Show this help message." _E
		"  --color " _I "Wrap matches with escapes to highlight them in red.\n"
		"          The use of color may be controlled as follows:" _E
		"  --color=auto   " _I "When the output is a TTY. Same as above." _E
		"  --color=never  " _I "Never. Overrides a previous option." _E
		"  --color-always " _I "Regardless of whether the output is a TTY." _E
		"\n"
		" Supported regex syntax:\n"
		"  [abc]   " _I "Match one of a set of characters." _E
		"  [a-z]   " _I "Match one from a range of characters." _E
		"  (abc)   " _I "Match a group.\n"
		"          This implementation does not support repeating a group match,\n"
		"          so groups do not really do anything." _E
		"  .       " _I "Match any single character." _E
		"  ^       " _I "Match the start of the line." _E
		"  $       " _I "Match the end of the line." _E
		"\n"
		" Modifiers (can be combined with [], ., and single characters):\n"
		"  *       " _I "Match any number of occurances" _E
		"  \\?      " _I "Match zero or one time" _E
		"  \\+      " _I "Match at least one occurance" _E
		"\n"
		" Some characters can be escaped in the pattern with \\.\n"
		" The regex engine is not Unicode-aware.\n",
		argv[0]);
#undef _I
#undef _E
	return 2;
}

int main(int argc, char ** argv) {
	int opt;
	while ((opt = getopt(argc, argv, "?hivqocFxlnsA:B:C:-:")) != -1) {
		switch (opt) {
			case 'h':
			case '?':
				return usage(argv);
			case 'i':
				ignorecase = 1;
				break;
			case 'v':
				invert = 1;
				break;
			case 'q':
				quiet = 1;
				break;
			case 'o':
				only_matching = 1;
				break;
			case 'c':
				counts = 1;
				break;
			case 'F':
				is_fgrep = 1;
				break;
			case 'x':
				whole_lines = 1;
				break;
			case 'l':
				list_files = 1;
				break;
			case 'n':
				line_numbers = 1;
				break;
			case 's':
				suppress_errors = 1;
				break;
			case 'A': {
				char * end = NULL;
				print_after = strtoul(optarg,&end,10);
				if (*end) return usage(argv);
				break;
			}
			case 'B': {
				char * end = NULL;
				print_before = strtoul(optarg,&end,10);
				if (*end) return usage(argv);
				break;
			}
			case 'C': {
				char * end = NULL;
				print_after = print_before = strtoul(optarg,&end,10);
				if (*end) return usage(argv);
				break;
			}
			case '-':
				if (!strcmp(optarg,"help")) {
					return usage(argv);
				} else if (!strcmp(optarg,"color=never")) {
					use_color = 0; /* Overrides previous instances of conflicting option */
				} else if (!strcmp(optarg,"color") || !strcmp(optarg,"color=auto")) {
					use_color = 1; /* treat 1 as maybe */
				} else if (!strcmp(optarg,"color=always")) {
					use_color = 2;
				}
				break;
		}
	}

	/* Detect if we were invoke as 'fgrep'. */
	if (!strcmp(basename(argv[0]),"fgrep")) {
		is_fgrep = 1;
	}

	/* Require at least a PATTERN argument. */
	if (optind == argc) return usage(argv);
	char * needle = argv[optind++];

	int ret = 1; /* Normal exit status: 0 if something matched, 1 if not. */
	int err = 0; /* Whether an error was encountered that should override the exit status to 2. */

	/* We show additional messages for detected binaries only if we are on a TTY,
	 * and "auto" color mode should only activate if we are on a TTY. */
	int is_tty = isatty(STDOUT_FILENO);
	if (!is_tty && use_color == 1) use_color = 0;

	/* When multiple file arguments are provided, include the names of files as a prefix to matches. */
	int showFilenames = (optind + 1 < argc);

	/* This buffer is allocated by 'getline' and reused between files. */
	char * buf = NULL;
	size_t avail = 0;

	do {
		FILE * input = stdin;
		ssize_t lineLength = 0;
		int count = 0;    /* Count of matching lines. */
		int isbinary = 0; /* If we have detected any NUL characters in the input. */
		int ln = 0;       /* Current line number. */
		size_t after = 0; /* How mines non-matching lines to print anyway */

		/* If there are file arguments, use them, but treat - as standard input. */
		if (optind < argc && strcmp(argv[optind],"-")) {
			input = fopen(argv[optind], "r");
			if (!input) {
				if (!suppress_errors) fprintf(stderr, "%s: %s: %s\n", argv[0], argv[optind], strerror(errno));
				/* We continue to read other arguments but note this error to exit with 2 later. */
				err = 1;
				optind++;
				continue;
			}
		}

		/* POSIX says we should print '(standard input)' instead of -, so make it happy. */
		const char * filename = input == stdin ? "(standard input)" : argv[optind];

		/* Clear context */
		dump_context(0);

		while ((lineLength = getline(&buf, &avail, input)) >= 0) {
			ln++;

			/* Ignore any trailing line feed. */
			if (lineLength && buf[lineLength-1] == '\n') {
				lineLength--;
			}

			/* Prepare Line for regex engine. */
			struct Line line = {
				lineLength,
				buf
			};

			/* Scan for any NUL characters in this line to see if it is a binary. */
			if (!isbinary) {
				for (ssize_t i = 0; i < lineLength; ++i) {
					if (buf[i] == '\0') {
						isbinary = 1;
						break;
					}
				}
			}

			/* isbinary can change whenever we hit a binary line, so recheck each time */
			int print_context = !quiet && !isbinary && !list_files && !counts;

			if (!isbinary && print_before) {
				struct ContextLine * ctx = malloc(sizeof(struct ContextLine));
				ctx->next = context_head;
				ctx->prev = NULL;
				ctx->len = lineLength;
				ctx->buf = strdup(buf); /* not binary, so this should be fine */
				ctx->filename = showFilenames ? filename : NULL;
				ctx->line = line_numbers ? ln : 0;
				if (context_head) {
					context_head->prev = ctx;
				}
				context_head = ctx;

				if (!context_len) {
					context_tail = ctx;
					context_len++;
				} else if (context_len == print_before + 1) {
					struct ContextLine * new_tail = context_tail->prev;
					new_tail->next = NULL;
					free(context_tail->buf);
					free(context_tail);
					context_tail = new_tail;
					context_overflowed = 1;
				} else {
					context_len++;
				}
			}

			if (!invert) {
				int lastMatch = 0; /* End of the last match. */
				int matched = 0;   /* Whether this line has matched yet. Useful when a degenerate match means lastMatch is still 0. */
				for (int j = 0; j <= lineLength;) {
					int len;
					if (subsearch_matches(&line, j, needle, &len)) {
						if (whole_lines && len != lineLength) break;

						/* If quiet gets a match at all, ony any file, we immediately exit with 0,
						 * even ignoring any errors we may have printed about previously. */
						if (quiet) return 0;

						/* Increment count of matching lines only if this is the fist match */
						if (!matched) {
							if (print_context && context_overflowed && (count || ret == 0)) print_dash("--",1);
							if (print_before && print_context) dump_context(1);

							count++;
						}

						/* If anything matched, return code is 0 except for an error. */
						ret = 0;
						matched = 1;

						/* If we are just listing matching files, we're done on the first match. */
						if (list_files) goto _done;

						/* If we're counting matching lines, we're done with this line. */
						if (counts) break;

						/* If this is a binary file, and we are not counting matched lines,
						 * we're done here, similar to listing. */
						if (isbinary) goto _done;

						if (!len && only_matching) {
							/* Don't try to print a degenerate match here. */
							lastMatch = j = j + 1;
							continue;
						}

						if (only_matching || !lastMatch) {
							/* Prefix this match result as needed. For -o, every match
							 * is prefixed. Otherwise, we only print the prefix before
							 * the first match. */
							if (showFilenames) print_filename(filename, ":");
							if (line_numbers) print_line_number(ln, ":");
						}

						if (only_matching) {
							fprintf(stdout, "%.*s\n", len, buf + j);
						} else {
							/* Print from the end of the previous match up to the end of
							 * the current match, with color when enabled. */
							fprintf(stdout, "%.*s%s%.*s%s",
								j - lastMatch,
								buf + lastMatch,
								use_color ? "\033[1;31m" : "",
								len,
								buf + j,
								use_color ? "\033[0m" : "");
						}

						/* Update lastMatch to point to the end of this match. */
						lastMatch = j + len;

						/* Advance to that point, ensuring we skip to the next character
						 * if this match was degenerate. */
						j = lastMatch + !len;
					} else {
						j++;
					}

					/* When matching whole lines, we only check from index 0, so break now. */
					if (whole_lines) break;
				}

				/* If we are counting matches, don't print anything and go to next line. */
				if (counts) continue;

				if (print_after && print_context) {
					if (matched) {
						after = print_after;
						if (!print_before) context_overflowed = 0;
					} else {
						if (after > 0) {
							dump_context(0);
							if (showFilenames) print_filename(filename, "-");
							if (line_numbers) print_line_number(ln, "-");
							fprintf(stdout, "%.*s\n", (int)lineLength, buf);
							after--;
						} else if (!print_before) {
							context_overflowed = 1;
						}
					}
				}

				/* If we weren't printing just the matching text and we had a match, there
				 * may be more of the line left to print, and we must also print a line feed
				 * ourselves (we ignore any line feed in the actual line, and we print one
				 * even if the line didn't actually end in one). */
				if (!only_matching && matched) fprintf(stdout, "%.*s\n", (int)(lineLength - lastMatch), buf + lastMatch);
			} else {

				/* The inverse matching case is a lot simpler as we don't have to deal with
				 * coloring submatches. Try the pattern at every point in the line, and if
				 * it ever matches, reject it. When un-matching whole lines, check only
				 * from index 0 and only reject if the match length is the same as the line. */
				int matched = 0;
				for (int j = 0; j <= lineLength; ++j) {
					int len;
					if (subsearch_matches(&line, j, needle, &len)) {
						if (whole_lines && len != lineLength) break;
						matched = 1;
						break;
					}
					if (whole_lines) break;
				}

				/* Do nothing on a matched line. */
				if (matched) {
					if (print_after && print_context) {
						if (after > 0) {
							dump_context(0);
							if (showFilenames) print_filename(filename, "-");
							if (line_numbers) print_line_number(ln, "-");
							fprintf(stdout, "%.*s\n", (int)lineLength, buf);
							after--;
						} else if (!print_before) {
							context_overflowed = 1;
						}
					}
					continue;
				}

				/* In quiet mode, if anything un-matched, immediately exit with 0,
				 * ignoring all the other arguments. */
				if (quiet) return 0;

				if (print_context) {
					if (context_overflowed && count) print_dash("--",1);
					if (print_before) dump_context(1);
					if (print_after) after = print_after;
					if (!print_before) context_overflowed = 0;
				}

				/* Otherwise any un-matched line sets our return status to 0 like any
				 * matched line does in the non-inverse case. */
				ret = 0;

				/* Count un-matched lines. */
				count++;

				/* If just listing file names, we're done with this file. */
				if (list_files) goto _done;

				/* If we're counting un-matching lines, we're done with this one. */
				if (counts) continue;

				/* And again, same as above, if this was a binary and we weren't
				 * counting un-matched 'lines', we're now done with this file. */
				if (isbinary) goto _done;

				/* -ov is a weird combo, but don't print anything. */
				if (only_matching) continue;

				/* Print relevant prefixes. */
				if (showFilenames) print_filename(filename, ":");
				if (line_numbers) print_line_number(ln, ":");

				/* And finally, print the un-matched line with a line feed. */
				fprintf(stdout, "%.*s\n", (int)lineLength, buf);
			}
		}

		if (lineLength < 0 && ferror(input)) {
			if (!suppress_errors) fprintf(stderr, "%s: %s: %s\n", argv[0], filename, strerror(errno));
			err = 1;
			optind++;
			continue;
		}

_done: (void)0;
		if (input != stdin) fclose(input);

		if (!quiet) {
			if (list_files) {
				if (count) puts(filename);
			} else if (counts) {
				if (showFilenames) fprintf(stdout, "%s:", filename);
				fprintf(stdout, "%d\n", count);
			} else if (count && isbinary && is_tty) {
				fprintf(stdout, "%s: %s: binary file matches\n", argv[0], filename);
			}
		}

		optind++;
	} while (optind < argc);

	return err ? 2 : ret;
}

